decorator_allowed_groups/models/decorator.py

```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2021-Today: GRAP (<http://www.grap.coop/>)
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api
+from odoo.exceptions import AccessError
+
+
+def allowed_groups(*group_xml_ids):
+    """ Return a decorator that specifies group(s)
+        to which the user must belong in order to perform
+        this function.
+        - if the user does not belong to any group,
+        an AccessError will be raised if the function is called.
+        - Also in the generated views, the according button
+        will be hidden if the user doesn't belong to the group(s).
+
+            @api.allowed_groups(
+                "purchase.group_purchase_manager",
+                "sale.group_sale_manager"
+            )
+            def my_secure_action(self):
+                pass
+    """
+
+    def decorator(method):
+        def secure_method(*args, **kwargs):
+            _self = args[0]
+            # Check if current user is member of correct groups
+            if any([
+                _self.env.user.has_group(group_xml_id)
+                for group_xml_id in group_xml_ids
+            ]):
+                # If it's OK, return the original method
+                return method(*args, **kwargs)
+            else:
+                # If it's KO, raise an error.
+                # We raise a technical message (with function name
+                # and xml_ids of the groups, because this message
+                # will be raised only in XML-RPC call.
+                raise AccessError(_(
+                    "To execute the function {}, you should be member"
+                    " of one of the following groups:\n {}".format(
+                        method,
+                        ', '.join(group_xml_ids))
```

Comment:
I believe applying `format` on user controlled input (such as a translatable string) has security implications (except that `format` is applied here to the translation source instead of the translation value but that is another thing).

Reply:
> applied here to the translation source instead of the translation value but that is another thing

Thanks ! Fixed.

> applying format on user controlled input (such as a translatable string) has security implications

I didn't know. I replaced by ``%s``. Is it better ? ;-)

Reply:
Yes, see https://github.com/OCA/pylint-odoo/issues/302